Common/src/main/java/com/blamejared/crafttweaker/api/util/sequence/Sequence.java

```diff
@@ -0,0 +1,99 @@
+package com.blamejared.crafttweaker.api.util.sequence;
+
+import com.blamejared.crafttweaker.api.annotation.ZenRegister;
+import com.blamejared.crafttweaker.api.util.sequence.task.ISequenceTask;
+import com.blamejared.crafttweaker_annotations.annotations.Document;
+import org.openzen.zencode.java.ZenCodeType;
+
+import java.util.LinkedList;
+import java.util.function.Supplier;
+
+/**
+ * A sequence is a series of tasks that run after each other when the actor ticks.
+ *
+ * <p>For example, when a level ticks.</p>
+ *
+ * @param <T>
+ * @param <U>
+ */
+@ZenRegister
+@Document("vanilla/api/util/sequence/Sequence")
+@ZenCodeType.Name("crafttweaker.api.util.sequence.Sequence")
+public class Sequence<T, U> {
+    
+    private final Supplier<T> actor;
+    private final LinkedList<ISequenceTask<T, U>> timeline;
+    private boolean stopped;
+    private final SequenceContext<T, U> context;
+    
+    public Sequence(Supplier<T> actor, U data, LinkedList<ISequenceTask<T, U>> timeline) {
+        
+        this.actor = actor;
+        this.timeline = timeline;
+        this.stopped = false;
+        this.context = new SequenceContext<>(this, data);
+    }
+    
+    /**
+     * Ticks this sequence.
+     */
+    @ZenCodeType.Method
+    public void tick() {
+        
+        if(timeline.isEmpty()) {
+            return;
+        }
+        ISequenceTask<T, U> task = timeline.peek();
+        
+        task.tick(actor.get(), context);
+        if(task.isComplete(actor.get(), context)) {
+            timeline.remove();
+        }
+    }
```

Comment:
So it will only ever do 1 task per tick, is that the way we want it to be?
Like, let's say I have
```
...sequenceBuilder
.sleepUntil((level) => level.rains)
.sleepUntil((level) => level.isDay)
.sleepUntil((level) => level.blockAt(5, 5, 5) === <block:minecraft:bedrock>.defaultState)
.sleepUntil((level) => level.canSeeTheSky(5, 5, 5))
.run((level) => println("This chunk of bedrock finally saw the rain at day!");
```

Let's say it is raining at day and the block at `5,5,5` is indeed bedrock that can see the sky.
Would it take 5 world ticks for this to be executed?

Reply:
Yes, the correct way to do this would be like:
```zenscript
...sequenceBuilder
.sleepUntil((level) => level.rains && level.isDay && level.blockAt(5, 5, 5) === <block:minecraft:bedrock>.defaultState && level.canSeeTheSky(5, 5, 5))
.run((level) => println("This chunk of bedrock finally saw the rain at day!");
```

We can add batched calls at a later point if it comes up